design/operator-architecture.asciidoc

```diff
@@ -0,0 +1,154 @@
+|Service Infrastructure Admin |Full Lifecycle | Recover Keycloak from a database backup| No
+|Service Infrastructure Admin |Deep Insights | Monitor Keycloak health| Yes
+|Service Infrastructure Admin |Deep Insights | Monitor Keycloak performance| Yes
+
+|Realm Admin |Basic Install | Provision new Realm, Client and User| Yes
+|Realm Admin |Basic Install | Secure an application| Yes
+|Realm Admin |Basic Install | Access Admin Console| Yes
+|Realm Admin |Basic Install | Provide transparent secure proxy (Gatekeeper) for third party applications| Yes
+|Realm Admin |Seamless Upgrades | Zero downtime upgrades| No
+|Realm Admin |Full Lifecycle | Backup created resources periodically| Yes
+|Realm Admin |Deep Insights | Analyze usage and performance metrics| Partially
+|Realm Admin |Autopilot | Scale Keycloak automatically depending on traffic| No
+
+|Secured Application |Basic Install | Accessing Keycloak using a secured connection| Yes
+|Secured Application |Seamless upgrades | Accessing Keycloak during the upgrade process| No
+|Secured Application |Full Lifecycle | Keep session secured even when facing a disaster| Partially
+|Secured Application |Autopilot | Provide stable response time despite the load| No
+|====
+
+## History
+
+Keycloak Operator was re-designed from the scratch by a small team withing RHMI and RHSSO groups. The implementation plan was described in the link:operator.md[Initial Operator Design Document]. At the end of a 3 sprint cycle, the team delivered a Level 4 Operator, that was pushed into the https://operatorhub.io/operator/keycloak-operator[Operatorhub]. Once the initial implementation was done, the Operator was handed over to the Keycloak/RHSSO Team.
+
+After maintaining the operator for several months, it became necessary to create a High Level Design (and a long-term vision) how all Keycloak components work together.
+
+## Layered architecture
+
+Keycloak ecosystem consists of 3 layers:
+
+* https://github.com/keycloak/keycloak-operator[Keycloak Operator]
+* https://github.com/keycloak/keycloak-containers/tree/master/server[Keycloak Container Image]
+* https://github.com/keycloak/keycloak[Keycloak Server]
+
+image::img/Keycloak-operator-layers.png[Keycloak ecosystem layers]
+
+Keycloak Operator uses Keycloak Container image, which in turn uses Keycloak Server bits. The goal is to implement features in the lowest possible layer, so that they get inherited by the all upper layers.
+
+.Feature implementation example
+****
+Let's consider a simple feature - adding JSON formatted logs based on an environmental variable.
+
+This feature requires touching all three layers:
+
+1. Keycloak Server needs to provide a JSON log based formatter.
+2. Keycloak Container needs to expose an environment variable, that could be switching on to use JSON formatting
+3. Keycloak Operator needs to expose proper configuration on Custom Resource level enabling JSON log formatting.
```

Comment:
Shouldn't the Operator actually be opinionated here and just use JSON logging since that is the recommended on OpenShift/Kubernetes to have a common/parsable logging format?